apis/apikeys/v1alpha1/apikey_generated.go

```diff
@@ -0,0 +1,179 @@
+// Copyright 2024 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package v1alpha1
+
+import (
+	"github.com/GoogleCloudPlatform/k8s-config-connector/apis/refs/v1beta1"
+)
+
+// +kcc:proto=google.api.apikeys.v2.AndroidApplication
+type AndroidApplication struct {
+	// The SHA1 fingerprint of the application. For example, both sha1 formats are
+	//  acceptable : DA:39:A3:EE:5E:6B:4B:0D:32:55:BF:EF:95:60:18:90:AF:D8:07:09 or
+	//  DA39A3EE5E6B4B0D3255BFEF95601890AFD80709.
+	//  Output format is the latter.
+	// +required
+	Sha1Fingerprint *string `json:"sha1Fingerprint,omitempty"`
```

Comment:
So `+required` works now, after https://github.com/kubernetes-sigs/controller-tools/pull/944

We don't yet validate the input format (though I would expect GCP to reject it if invalid, so the user will still get a reasonable error, we shouldn't fail silently)

Because the input and output formats are potentially different, we'll need to be careful here, we can't just do a naive diff to determine if we need to re-reconcile.

Reply:
> So `+required` works now, after [kubernetes-sigs/controller-tools#944](https://github.com/kubernetes-sigs/controller-tools/pull/944)
> 

Ah! This is so fresh and good news!

> We don't yet validate the input format (though I would expect GCP to reject it if invalid, so the user will still get a reasonable error, we shouldn't fail silently)
> 

Agree. I think I want to have basic and general validation on CRD, and have more server specific/usage related validation from the GCP server.   

> Because the input and output formats are potentially different, we'll need to be careful here, we can't just do a naive diff to determine if we need to re-reconcile.

Yes. This makes a lot of sense.